pkg/audit/audit.go

```diff
@@ -0,0 +1,139 @@
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	k8stypes "k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	contentTypeJSON = "application/json"
+)
+
+const (
+	levelNull = iota
+	levelMetadata
+	levelRequest
+	levelRequestResponse
+)
+
+var (
+	bodyMethods = map[string]bool{
+		http.MethodPut:  true,
+		http.MethodPost: true,
+	}
+)
+
+type auditLog struct {
+	log     *log
+	writer  *LogWriter
+	reqBody []byte
+}
+
+type log struct {
+	AuditID           k8stypes.UID `json:"auditID,omitempty"`
+	RequestURI        string       `json:"requestURI,omitempty"`
+	User              *userInfo    `json:"user,omitempty"`
+	Verb              string       `json:"verb,omitempty"`
+	SourceIP          string       `json:"sourceIP,omitempty"`
+	RequestTimestamp  string       `json:"requestTimestamp,omitempty"`
+	ResponseTimestamp string       `json:"responseTimestamp,omitempty"`
+	ResponseCode      int          `json:"responseCode,omitempty"`
+	RequestHeader     http.Header  `json:"requestHeader,omitempty"`
+	ResponseHeader    http.Header  `json:"responseHeader,omitempty"`
+	RequestBody       []byte       `json:"requestBody,omitempty"`
+	ResponseBody      []byte       `json:"responseBody,omitempty"`
+}
+
+type userInfo struct {
+	Name  string `json:"name,omitempty"`
+	Group string `json:"group,omitempty"`
+}
+
+func new(writer *LogWriter, req *http.Request) (*auditLog, error) {
+	auditLog := &auditLog{
+		writer: writer,
+		log: &log{
+			AuditID:          k8stypes.UID(uuid.NewRandom().String()),
+			RequestURI:       req.RequestURI,
+			Verb:             req.Method,
+			SourceIP:         getSourceIPs(req.RemoteAddr),
+			RequestTimestamp: time.Now().Format(time.RFC3339),
+		},
+	}
+
+	contentType := req.Header.Get("Content-Type")
+	if writer.Level >= levelRequest && bodyMethods[req.Method] && contentType == contentTypeJSON {
+		reqBody, err := readBodyWithoutLosingContent(req)
+		if err != nil {
+			return nil, err
+		}
+		auditLog.reqBody = reqBody
+	}
+	return auditLog, nil
+}
+
+func (a *auditLog) write(reqHeaders, resHeaders http.Header, resCode int, resBody []byte) error {
+	userInfo := &userInfo{
+		Name:  reqHeaders.Get("Impersonate-User"),
+		Group: reqHeaders.Get("Impersonate-Group"),
+	}
+	a.log.User = userInfo
+	a.log.ResponseTimestamp = time.Now().Format(time.RFC3339)
+	a.log.RequestHeader = reqHeaders
+	a.log.ResponseHeader = resHeaders
+	a.log.ResponseCode = resCode
+
+	var buffer bytes.Buffer
+	alByte, err := json.Marshal(a.log)
+	if err != nil {
+		return err
+	}
+
+	buffer.Write(bytes.TrimSuffix(alByte, []byte("}")))
+
+	if a.writer.Level >= levelRequest && a.reqBody != nil {
```

Comment:
This shouldn't be required as the []byte fields are already encoded from the Marshal call before.

Reply:
Do you mean ``` a.reqBody != nil ```? because in some case like not content-type not json the body will be nil.